File: modules/CalicoParser/android/src/main/java/expo/module/calicoparser/epub/Epub.kt
package expo.module.calicoparser.epub

import android.content.Context
import android.net.Uri
import android.util.Log

class EpubParser {
    private val zip = Zip()
    private val xml = XML()
    private val normalize = Normalize()

    fun parseChapter(context: Context, uri: Uri, chapterPath: String): String {
        return zip.parseFile(context, uri, chapterPath)
    }

    fun parseChunk(context: Context, uri: Uri, filePaths: List<String>): Map<String, String> {
        return zip.parseMultipleFiles(context, uri, filePaths)
    }

    fun chunkEpub(context: Context, uri: Uri): Map<String, List<String>> {
        val containerxml = zip.parseFile(context, uri, "META-INF/container.xml")
        val rootfile = xml.parseAttribute(containerxml, "rootfile", "full-path")
        val rootfileContent = zip.parseFile(context, uri, rootfile)
        val chunks = normalize.getChunks(context, uri, rootfileContent)
        return chunks
    }

    fun importMetadata(context: Context, uri: Uri): Map<String, Any?> {
        val (containerXml, rootfileContent) = zip.getBaseFiles(context, uri)

        if (containerXml == null || rootfileContent == null) {
            throw Exception("Failed to find containerxml or rootfilecontent")
        }

        val metadata = normalize.buildMetadata(rootfileContent)
        return metadata
    }

    fun cleanCache() {
        zip.clean()
    }
}

File: lib/epub.ts
import CalicoParser from '@/modules/CalicoParser';
import { Metadata } from '@/modules/CalicoParser';

const loadChapter = async (uri: string, chapterPath: string): Promise<String> => {
    let result = '';
    try {
        result = await CalicoParser.parseChapter(uri, chapterPath);
    } catch (e) {
        console.error(e);
    }
    return result;
};

const importMetadata = async (uri: string): Promise<Metadata | null> => {
    let result = null;
    try {
        result = await CalicoParser.importMetadata(uri);
    } catch (e) {
        console.error(e);
    }
    return result;
};

const parseChunk = async (uri: string, chapterPaths: Array<string>): Promise<Record<string, string>> => {
    const result = await CalicoParser.parseChunk(uri, chapterPaths);
    return result;
};

const getChunks = async (uri: string): Promise<Record<string, string[]>> => {
    const result = await CalicoParser.chunkEpub(uri);
    return result;
};

const cleanCache = async (): Promise<void> => {
    CalicoParser.clean();
};

export { loadChapter, importMetadata, getChunks, parseChunk, cleanCache };
